### PR TITLE
lower: write out logic to convert TIR types into IR types

### PR DESCRIPTION
**Pull request by feds01**
This patch kicks off the migration of the compiler from the old typechecker to the new typechecker. This patch implements conversion logic between the TIR types into the IR types.

### CODE REVIEW

File: compiler/hash-lower/src/ty/new.rs [hunk 1]
```diff
@@ -0,0 +1,263 @@
...
+                                let fields = fields
+                                    .iter()
+                                    .map(|field| {
+                                        let ty = self.ty_id_from_tir_ty(field.ty);
+                                        let name = self
+                                            .get_symbol(field.name)
+                                            .name
+                                            .unwrap_or(IDENTS.underscore);
+                                        // @@Verify: do we need to store default parameter
+                                        // information here?
```
Comment (kontheocharis) [inline compiler/hash-lower/src/ty/new.rs:203]: no

File: compiler/hash-lower/src/ty/new.rs [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+                                    }
+                                }
+                                // We don't implement bigints yet
+                                _ => unimplemented!(),
+                            }
+                        }
+                    }
+
+                    // @@Verify: does `str` now imply that it is a `&str`, or should we create a
+                    // `&str` type here.
```
Comment (kontheocharis) [inline compiler/hash-lower/src/ty/new.rs:253]: Probably the latter, maybe better to do this once TC deals with it?

**Review by kontheocharis — APPROVED** [review]

File: compiler/hash-lower/src/ty/new.rs [hunk 3]
```diff
@@ -0,0 +1,263 @@
...
+                let params = self.stores().params().map_fast(*params, |params| {
+                    self.lcx.tls().create_from_iter(
+                        params.iter().map(|param| self.ty_id_from_tir_ty(param.ty)),
+                    )
+                });
+
+                let return_ty = self.ty_id_from_tir_ty(*return_ty);
+
+                // @@Todo: functions should have a name if possible, otherwise they default
+                // to an underscore.
```
Comment (kontheocharis) [inline compiler/hash-lower/src/ty/new.rs:127]: What, like function types?

Comment (feds01) [reply compiler/hash-lower/src/ty/new.rs:127]: Yes, the function type/definition should store associated attributes, one of them being a name

File: compiler/hash-lower/src/ty/new.rs (hunk 2 above)
Comment (feds01) [reply compiler/hash-lower/src/ty/new.rs:253]: As in that it verifies that the `str` does not appear in a type position (except generic parameters)?

Comment (kontheocharis) [reply compiler/hash-lower/src/ty/new.rs:253]: Yeah, surely that should be part of the layout logic?

File: compiler/hash-lower/src/ty/new.rs (hunk 3 above)
Comment (kontheocharis) [reply compiler/hash-lower/src/ty/new.rs:127]: A function type is different from a function definition.

**Pull request merged**